service/controller/clusterapi/v29/key/cluster.go

```diff
@@ -36,11 +36,19 @@ const (
 	KubernetesSecurePort = 443
 )
 
+// MaximumNumberOfAZsInCluster defines the current limitation with Node Pools
+// implementation. Biggest limitation behind this is current IPAM
+// implementation that restricts network sizes. Another related problem is
+// restrictions in AWS resource structure.
+const MaximumNumberOfAZsInCluster = 4
```

Comment:
We should not have this as a magic number somewhere in the operator IMO. This will eventually cause confusion because someone can't find the reason why he/she can't add more node pools.

Reply:
Or when do we want to change this?

Reply:
We must have this in this operator as it is an important piece of whole design of node pools infrastructure and behavior of dynamic AZs. I expect this to change when we can solve `Advanced IPAM` roadmap story and somewhere in the future possibly also refactor the TCCP structure further.

> because someone can't find the reason why he/she can't add more node pools.

Asynchronous CR reconciliation is not a design where one can provide reasonable feedback to end user. Somewhere in the future we might want to submit an event for these kind of cases, but that requires better error structuring. For end user UX there must be measures in place in `api`, `gsctl` and `happa`.